### PR TITLE
Add isWeaponUsed scripting function

### PR DESCRIPTION
**Pull request by DahBlount**
Requested by Axem.

Allows a weapon to be paged in via script so if a weapon has special qualities through scripting, say, upgrades given a player state, the weapons can be paged in and rendered correctly.

### CODE REVIEW

**Comment by asarium** [comment]
IMO it would be better to either split the functionality into two functions where one queries the value and one pages the weapon in or it could use a variable with which a script could mark a weapon as used which would also page in the resources of the weapon.

**Comment by DahBlount** [comment]
This was designed to be functionally like the isModelLoaded function for ship handles at the request of Axem.

The issue with the script marking it as used after the fact is that the weapons only get paged in during level paging, not during mission run time. So you have to force the weapons to get paged in and mark them used simultaneously.

**Comment by asarium** [comment]
Ok, then the two function approach seems to be the better solution. Grouping two functions together into one function is a bad design choice since the function does two things but the function name only implies one thing. I think two functions would be better in this situation.

**Comment by DahBlount** [comment]
Alright. Here are the changes.

File: code/scripting/lua.cpp [hunk 1]
```diff
@@ -4637,6 +4637,40 @@ ADE_FUNC(isBeam, l_Weaponclass, NULL, "Return true if the weapon is a beam", "bo
 		return ADE_RETURN_FALSE;
 }
 
+// Checks if a weapon has been paged in (counted as used)
+ADE_FUNC(isWeaponUsed, l_Weaponclass, NULL, "Return true is the weapon is paged in, optionally paging in the weapon.", "boolean", "true if the weapon is paged in, pages in the weapon if otherwise")
```
Comment (MageKing17) [inline code/scripting/lua.cpp]: This description would seem to be wrong.

File: code/scripting/lua.cpp [hunk 2]
```diff
@@ -4637,6 +4637,40 @@ ADE_FUNC(isBeam, l_Weaponclass, NULL, "Return true if the weapon is a beam", "bo
...
+{
+	int idx;
```
Comment (MageKing17) [inline code/scripting/lua.cpp]: Leaving `idx` uninitialized is probably not a good idea (we've had plenty of problems with uninitialized variables in the Lua interface already).

File: code/scripting/lua.cpp [hunk 3]
```diff
@@ -4637,6 +4637,40 @@ ADE_FUNC(isBeam, l_Weaponclass, NULL, "Return true if the weapon is a beam", "bo
...
+		return ADE_RETURN_FALSE;
+
+	if (!weapon_used(idx)) {
+		return ADE_RETURN_FALSE;
+	} else {
+		return ADE_RETURN_TRUE;
+	}
+}
+
+// Checks if a weapon has been paged in (counted as used)
```
Comment (MageKing17) [inline code/scripting/lua.cpp]: This comment is a duplicate of the one for the previous function.

File: code/scripting/lua.cpp [hunk 4]
```diff
@@ -4637,6 +4637,40 @@ ADE_FUNC(isBeam, l_Weaponclass, NULL, "Return true if the weapon is a beam", "bo
...
+ADE_FUNC(loadWeapon, l_Weaponclass, NULL, "Pages in a weapon. Returns True on success.", "boolean", "True if page in was successful, false otherwise.")
+{
+	int idx;
```
Comment (MageKing17) [inline code/scripting/lua.cpp]: See previous comment about uninitialized `idx`.

File: code/scripting/lua.cpp [hunk 5]
```diff
@@ -4637,6 +4637,40 @@ ADE_FUNC(isBeam, l_Weaponclass, NULL, "Return true if the weapon is a beam", "bo
 		return ADE_RETURN_FALSE;
 }
 
+// Checks if a weapon has been paged in (counted as used)
+ADE_FUNC(isWeaponUsed, l_Weaponclass, NULL, "Return true is the weapon is paged in, optionally paging in the weapon.", "boolean", "true if the weapon is paged in, pages in the weapon if otherwise")
+{
+	int idx;
+	if (!ade_get_args(L, "o|b", l_Weaponclass.Get(&idx)))
```
Comment (MageKing17) [inline code/scripting/lua.cpp]: No optional boolean argument anymore.

File: code/scripting/lua.cpp [hunk 6]
```diff
@@ -4637,6 +4637,40 @@ ADE_FUNC(isBeam, l_Weaponclass, NULL, "Return true if the weapon is a beam", "bo
...
+	} else {
+		return ADE_RETURN_TRUE;
+	}
+}
+
+// Checks if a weapon has been paged in (counted as used)
+ADE_FUNC(loadWeapon, l_Weaponclass, NULL, "Pages in a weapon. Returns True on success.", "boolean", "True if page in was successful, false otherwise.")
+{
+	int idx;
+	if (!ade_get_args(L, "o|b", l_Weaponclass.Get(&idx)))
```
Comment (MageKing17) [inline code/scripting/lua.cpp]: This function should never have had an optional boolean argument.

**Comment by DahBlount** [comment]
Applied the suggested changes.

**Pull request merged**